pkg/scaling/scale_handler.go

```diff
@@ -476,62 +492,78 @@ func (h *scaleHandler) GetScaledObjectMetrics(ctx context.Context, scaledObjectN
-						metricscollector.RecordScalerLatency(scaledObjectNamespace, scaledObject.Name, scalerName, triggerIndex, metricName, float64(latency))
+					var metrics []external_metrics.ExternalMetricValue
+
+					// if cache is defined for this scaler/metric, let's try to hit it first
+					metricsFoundInCache := false
+					if scalerConfig.TriggerUseCachedMetrics {
+						var metricsRecord metricscache.MetricsRecord
+						if metricsRecord, metricsFoundInCache = h.scaledObjectsMetricCache.ReadRecord(scaledObjectIdentifier, metricName); metricsFoundInCache {
+							logger.V(1).Info("Reading metrics from cache", "scaler", triggerName, "metricName", metricName, "metricsRecord", metricsRecord)
+							metrics = metricsRecord.Metric
+							err = metricsRecord.ScalerError
+						}
 					}
-					logger.V(1).Info("Getting metrics from scaler", "scaler", scalerName, "metricName", spec.External.Metric.Name, "metrics", metrics, "scalerError", err)
-				}
 
-				// check if we need to set a fallback
-				metrics, fallbackApplied, err := fallback.GetMetricsWithFallback(ctx, h.client, metrics, err, metricName, scaledObject, spec)
-				if fallbackApplied {
-					fallbackActive = true
-				}
-				if err != nil {
-					isScalerError = true
-					logger.Error(err, "error getting metric for scaler", "scaler", scalerName)
-				} else {
-					for _, metric := range metrics {
-						metricValue := metric.Value.AsApproximateFloat64()
-						metricscollector.RecordScalerMetric(scaledObjectNamespace, scaledObjectName, scalerName, triggerIndex, metric.MetricName, metricValue)
+					if !metricsFoundInCache {
+						var latency int64
+						metrics, _, latency, err = cache.GetMetricsAndActivityForScaler(ctx, triggerIndex, metricName)
+						if latency != -1 {
+							metricscollector.RecordScalerLatency(scaledObjectNamespace, scaledObject.Name, triggerName, triggerIndex, metricName, float64(latency))
+						}
+						logger.V(1).Info("Getting metrics from trigger", "trigger", triggerName, "metricName", metricName, "metrics", metrics, "scalerError", err)
 					}
-					matchingMetrics = append(matchingMetrics, metrics...)
-				}
-				metricscollector.RecordScalerError(scaledObjectNamespace, scaledObjectName, scalerName, triggerIndex, metricName, err)
+					result.metricName = metricName
+					result.triggerName = triggerName
+					result.triggerIndex = triggerIndex
+					result.metricSpec = spec
+					result.metrics = metrics
+					result.err = err
+					results <- result
+					wg.Done()
+				}(matchingMetricsChan, &wg, metricName, triggerIndex, scalerConfigs[triggerIndex], spec)
 			}
 		}
 	}
 
+	wg.Wait()
+	close(matchingMetricsChan)
+	for result := range matchingMetricsChan {
+		for key, value := range result.metricTriggerPair {
+			metricTriggerPairList[key] = value
+		}
+		// check if we need to set a fallback
+		metrics, fallbackActive, err := fallback.GetMetricsWithFallback(ctx, h.client, result.metrics, result.err, result.metricName, scaledObject, result.metricSpec)
```

Comment:
I realize this is how it was implemented but perhaps during some follow-up refactoring, we can see if instead of running a `PATCH` for each metric, we could first aggregate the status changes and then run just a single `PATCH` for the entire `ScaledObject`. Not critical API call optimization :)

Reply:
That's a really good point. What about opening an issue to track it and not blocking the feature for the release? (I can work in the refactor this weekend if we want to have it before the release)

Reply:
Let's do this post release, no need to rush on this, let's be careful :)

Reply:
Yeah, that's why I suggested the issue :) Internal refactors can be done and shipped at any moment

Reply:
https://github.com/kedacore/keda/issues/5359

Reply:
Agreed, post release seems like the right time for this refactor.